### PR TITLE
Allow ChannelCore implementors to unwrap NIOAny.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

When implementing a custom ChannelCore, you will probably need access
to the data inside a NIOAny. It should be possible to unwrap that.

Modifications:

Added an extension to ChannelCore to allow unwrapping a NIOAny.
Added @_versioned to almost all of NIOAny.

Result:

ChannelCore is implementable outside NIO

### CODE REVIEW

File: Sources/NIO/NIOAny.swift [hunk 1]
```diff
@@ -77,6 +77,7 @@ public struct NIOAny {
     /// Try unwrapping the wrapped message as `ByteBuffer`.
     ///
     /// returns: The wrapped `ByteBuffer` or `nil` if the wrapped message is not a `ByteBuffer`.
+    @_versioned
```
Comment (weissi) [inline Sources/NIO/NIOAny.swift]: @Lukasa why not `@_inlineable` as well?

Comment (Lukasa) [reply Sources/NIO/NIOAny.swift]: Not allowed on public methods according to my compiler.

Comment (Lukasa) [reply Sources/NIO/NIOAny.swift]: The current reason is that `storage` is `private`. Do we think that's worth changing?

Comment (weissi) [reply Sources/NIO/NIOAny.swift]: @Lukasa not for this particular one but possibly for the ones that take generics.

Comment (Lukasa) [reply Sources/NIO/NIOAny.swift]: Yeah, I'm a bit on the fence here. I suppose making the storage internal won't be too big a deal.

Comment (weissi) [reply Sources/NIO/NIOAny.swift]: me too tbh. I'll make it your call. If you make the storage internal, please underscore it

Comment (Lukasa) [reply Sources/NIO/NIOAny.swift]: Ok, made all of these `@_inlineable`.

File: Sources/NIO/Channel.swift [hunk 2]
```diff
@@ -244,6 +244,28 @@ public extension Channel {
     }
 }
 
+public extension ChannelCore {
+    /// Unwraps the given NIOAny as a specific concrete type.
```
Comment (normanmaurer) [inline Sources/NIO/Channel.swift]: back-ticks around `NIOAny`

Comment (Lukasa) [reply Sources/NIO/Channel.swift]: Applied.

**Comment by weissi** [comment]
test failure should be addressed by #326

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Pull request merged**